// filter-suggest/lib/index.js
'use strict';

exports.__esModule = true;
exports.TextField = exports.Input = exports.HelperText = undefined;

var _extends = Object.assign || function (target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i]; for (var key in source) { if (Object.prototype.hasOwnProperty.call(source, key)) { target[key] = source[key]; } } } return target; };

var _react = require('react');

var _react2 = _interopRequireDefault(_react);

var _propTypes = require('prop-types');

var _propTypes2 = _interopRequireDefault(_propTypes);

var _downshift = require('downshift');

var _downshift2 = _interopRequireDefault(_downshift);

var _reactTextField = require('@material/react-text-field');

var _reactTextField2 = _interopRequireDefault(_reactTextField);

var _reactList = require('@material/react-list');

var _reactList2 = _interopRequireDefault(_reactList);

var _circularProgress = require('@rmwc/circular-progress');

require('@rmwc/circular-progress/circular-progress.css');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var FilterSuggest = function FilterSuggest(_ref) {
  var className = _ref.className,
      errorMessage = _ref.errorMessage,
      inputValue = _ref.inputValue,
      items = _ref.items,
      label = _ref.label,
      leadingIcon = _ref.leadingIcon,
      loading = _ref.loading,
      maxSuggestions = _ref.maxSuggestions,
      menuClassName = _ref.menuClassName,
      onBlur = _ref.onBlur,
      onFocus = _ref.onFocus,
      _onInputValueChange = _ref.onInputValueChange,
      onSelect = _ref.onSelect,
      outlined = _ref.outlined,
      textFieldClassName = _ref.textFieldClassName;

  return _react2.default.createElement(
    _downshift2.default,
    {
      selectedItem: null,
      inputValue: inputValue,
      onInputValueChange: function onInputValueChange(x) {
        return _onInputValueChange(x || '');
      },
      itemToString: function itemToString(item) {
        return item ? item.primary : '';
      },
      defaultHighlightedIndex: 0,
      onStateChange: function onStateChange(changes, downshift) {
        if (changes.hasOwnProperty('selectedItem')) {
          if (changes.selectedItem.primary) {
            return onSelect(changes.selectedItem);
          }
          return downshift.openMenu();
        }
        if (changes.type === _downshift2.default.stateChangeTypes.changeInput) {
          if (downshift.highlightedIndex !== 0) {
            downshift.setHighlightedIndex(0);
          }
        }
      }
    },
    function (_ref2) {
      var getInputProps = _ref2.getInputProps,
          getItemProps = _ref2.getItemProps,
          getLabelProps = _ref2.getLabelProps,
          getMenuProps = _ref2.getMenuProps,
          isOpen = _ref2.isOpen,
          inputValue = _ref2.inputValue,
          highlightedIndex = _ref2.highlightedIndex,
          setHighlightedIndex = _ref2.setHighlightedIndex,
          selectedItem = _ref2.selectedItem;

      var classNames = [];
      if (className) {
        classNames.push(className);
      }
      if (outlined) {
        classNames.push('fs-outlined');
      }
      if (isOpen && items.length) {
        classNames.push('fs-open');
      }
      return _react2.default.createElement(
        'div',
        { className: classNames.join(' ') },
        _react2.default.createElement(
          _reactTextField2.default,
          {
            outlined: outlined,
            className: 'fs-search-text-field ' + (textFieldClassName || ''),
            helperText: errorMessage ? _react2.default.createElement(
              _reactTextField.HelperText,
              null,
              _react2.default.createElement(
                'span',
                { style: { color: 'red' } },
                errorMessage
              )
            ) : null,
            label: label,
            leadingIcon: leadingIcon ? leadingIcon : undefined,
            trailingIcon: loading ? _react2.default.createElement(_circularProgress.CircularProgress, null) : undefined
          },
          _react2.default.createElement(_reactTextField.Input, _extends({}, getInputProps(), {
            className: 'fs-search-input',
            type: 'search',
            'data-lpignore': true,
            onBlur: onBlur,
            onFocus: onFocus
          }))
        ),
        !errorMessage && isOpen && items.length > 0 ? _react2.default.createElement(
          'div',
          {
            style: { position: 'relative' }
          },
          _react2.default.createElement(
            'div',
            _extends({}, getMenuProps(), {
              className: 'fs-filter-menu ' + (menuClassName || '')
            }),
            _react2.default.createElement(
              _reactList2.default,
              { twoLine: true },
              items.slice(0, maxSuggestions).map(function (item, index) {
                return _react2.default.createElement(
                  _reactList.ListItem,
                  _extends({}, getItemProps({ item: item }), {
                    key: item.id,
                    selected: highlightedIndex === index,
                    activated: false
                  }),
                  item.icon ? _react2.default.createElement(_reactList.ListItemGraphic, { graphic: item.icon }) : _react2.default.createElement('span', null),
                  _react2.default.createElement(_reactList.ListItemText, { primaryText: item.primary, secondaryText: item.secondary || ' ' }),
                  _react2.default.createElement(_reactList.ListItemMeta, { meta: highlightedIndex === index ? 'Enter' : ' ' })
                );
              })
            )
          )
        ) : null
      );
    }
  );
};
FilterSuggest.propTypes = process.env.NODE_ENV !== "production" ? {
  className: _propTypes2.default.string,
  errorMessage: _propTypes2.default.string,
  inputValue: _propTypes2.default.string.isRequired,
  items: _propTypes2.default.arrayOf(_propTypes2.default.shape({
    id: _propTypes2.default.string.isRequired,
    icon: _propTypes2.default.element,
    primary: _propTypes2.default.string.isRequired,
    secondary: _propTypes2.default.string
  })).isRequired,
  label: _propTypes2.default.string,
  leadingIcon: _propTypes2.default.element,
  loading: _propTypes2.default.bool,
  maxSuggestions: _propTypes2.default.number,
  menuClassName: _propTypes2.default.string,
  onBlur: _propTypes2.default.func,
  onFocus: _propTypes2.default.func,
  onInputValueChange: _propTypes2.default.func.isRequired,
  onSelect: _propTypes2.default.func.isRequired,
  outlined: _propTypes2.default.bool,
  textFieldClassName: _propTypes2.default.string
} : {};
FilterSuggest.defaultProps = {
  label: 'Start typing...',
  maxSuggestions: 12
};

exports.default = FilterSuggest;
exports.HelperText = _reactTextField.HelperText;
exports.Input = _reactTextField.Input;
exports.TextField = _reactTextField2.default;
